File: typescript/src/generate-key-pair.ts
import {
  exportKey,
  generateKeyPair as generateX25519KeyPair,
  generateSignKeyPair as generateEd25519KeyPair
} from 'stedy'
import { KeyPair } from '../types'

const exportKeyPair = async ({
  publicKey,
  privateKey
}: KeyPair): Promise<KeyPair> => ({
  publicKey: await exportKey(publicKey),
  privateKey: await exportKey(privateKey)
})

export const generateEphemeralKeyPair = async () =>
  exportKeyPair(await generateX25519KeyPair())

export const generateKeyPair = async () =>
  exportKeyPair(await generateEd25519KeyPair())

File: cplusplus/src/generate_key_pair.cpp
#include "autograph/generate_key_pair.h"

#include "sodium.h"

bool generate_ephemeral_key_pair(unsigned char *private_key,
                                 unsigned char *public_key) {
  int result = crypto_box_keypair(public_key, private_key);
  return result == 0;
}

bool generate_identity_key_pair(unsigned char *private_key,
                                unsigned char *public_key) {
  int result = crypto_sign_keypair(public_key, private_key);
  return result == 0;
}
